indico/modules/oauth/provider_test.py

```diff
@@ -0,0 +1,169 @@
+#
+# Indico is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from datetime import datetime, timedelta
+from uuid import uuid4
+
+import pytest
+from flask import session
+from freezegun import freeze_time
+from mock import MagicMock
+from oauthlib.oauth2 import InvalidClientIdError
+from sqlalchemy.orm.exc import NoResultFound
+
+from indico.core.config import Config
+from indico.modules.oauth.provider import load_client, load_token, save_grant, save_token, DisabledClientIdError
+from indico.modules.oauth.models.applications import OAuthApplication
+from indico.modules.oauth.models.tokens import OAuthGrant
+
+pytest_plugins = 'indico.modules.oauth.testing.fixtures'
+
+
+@pytest.fixture
+def token_data():
+    return {'access_token': str(uuid4()),
+            'expires_in': 3600,
+            'refresh_token': '',
+            'scope': 'api'}
+
+
+@pytest.fixture
+def create_request(dummy_application, dummy_user):
+    def _create_request(implicit=False):
+        request = MagicMock()
+        request.grant_type = 'authorization_code' if not implicit else None
+        request.client.client_id = dummy_application.client_id
+        request.user = dummy_user
+        return request
+    return _create_request
+
+
+@pytest.fixture
+def dummy_request(create_request):
+    return create_request()
+
+
+def test_load_client(dummy_application):
+    assert load_client(dummy_application.client_id) == dummy_application
+
+
+def test_load_client_malformed_id():
+    with pytest.raises(InvalidClientIdError):
+        load_client('foobar')
+
+
+def test_load_client_disabled_app(dummy_application):
+    dummy_application.is_enabled = False
+    with pytest.raises(DisabledClientIdError):
+        load_client(dummy_application.client_id)
+
+
+@freeze_time(datetime.utcnow())
+@pytest.mark.usefixtures('request_context')
+def test_save_grant(mocker):
+    mocker.patch.object(OAuthGrant, 'save')
+    request = MagicMock()
+    request.scopes = 'api'
+    request.redirect_uri = 'http://localhost:5000'
+    client_id = str(uuid4())
+    code = {'code': 'foobar'}
+    ttl = Config.getInstance().getOAuthGrantTokenTTL()
+    expires = datetime.utcnow() + timedelta(seconds=ttl)
+    grant = save_grant(client_id, code, request)
+    assert grant.client_id == client_id
+    assert grant.code == code['code']
+    assert grant.redirect_uri == request.redirect_uri
+    assert grant.user == session.user
+    assert grant.scopes == request.scopes
+    assert grant.expires == expires
+    assert grant.save.called
+
+
+@pytest.mark.usefixtures('request_context')
+@pytest.mark.parametrize(('access_token'), (True, False))
```

Comment:
I know, thanks. I had more elements before and I simply forgot to remove the parenthesis.